rust_snuba/src/metrics/statsd.rs

```diff
@@ -0,0 +1,82 @@
+#![allow(dead_code)]
+
+use cadence::prelude::*;
+use cadence::{BufferedUdpMetricSink, MetricBuilder, MetricError, QueuingMetricSink, StatsdClient};
+use rust_arroyo::utils::metrics::Metrics as ArroyoMetrics;
+use std::collections::HashMap;
+use std::net::UdpSocket;
+
+#[derive(Debug)]
+pub struct StatsDBackend {
+    client: StatsdClient,
+}
+
+impl StatsDBackend {
+    pub fn new(host: &str, port: u16, prefix: &str, global_tags: HashMap<&str, &str>) -> Self {
+        let socket = UdpSocket::bind("0.0.0.0:0").unwrap();
+        socket.set_nonblocking(true).unwrap();
+        let sink_addr = (host, port);
+        let buffered = BufferedUdpMetricSink::from(sink_addr, socket).unwrap();
+        let queuing_sink = QueuingMetricSink::from(buffered);
+
+        let mut client_builder = StatsdClient::builder(prefix, queuing_sink);
+        for (k, v) in global_tags {
+            client_builder = client_builder.with_tag(k, v);
+        }
+
+        Self {
+            client: client_builder.build(),
+        }
+    }
+
+    fn send_with_tags<'t, T: cadence::Metric + From<String>>(
+        &self,
+        mut builder: MetricBuilder<'t, '_, T>,
+        tags: Option<HashMap<&'t str, &'t str>>,
+    ) -> Result<T, MetricError> {
+        if let Some(t) = tags {
+            for (key, value) in t {
+                builder = builder.with_tag(key, value);
+            }
+        }
+
+        builder.try_send()
+    }
+}
+
+impl ArroyoMetrics for StatsDBackend {
+    fn increment(&self, key: &str, value: Option<i64>, tags: Option<HashMap<&str, &str>>) {
+        if let Err(e) =
+            self.send_with_tags(self.client.count_with_tags(key, value.unwrap_or(1)), tags)
+        {
+            log::error!("Error sending metric: {}", e);
```

Comment:
I would not send sentry errors for individual metric failures, which is what I think this might do?

Reply:
It should but suspect it won't yet since we don't have a sentry logging integration. Changed to debug as discussed elsewhere.